File: web/dart/blocks/version-manager.dart
part of NetTango;

class VersionManager {
  static final VERSION = 2;

  static void updateWorkspace(Map<String, Object> json) {
    int version = json.containsKey("version") ? json["version"] : 0;

    if (version > VERSION) {
      throw "Somehow the given model version ($version) is greater than the supported NetTango version ($VERSION).";
    }

    if (version < 1) {
      updateVersion1(json);
    }

    if (version < 2) {
      updateVersion2(json);
    }

    json["version"] = VERSION;
  }

  static int addIdsToParamsAndProps(int attributeId, Map<String, Object> b) {
    if (b.containsKey("params") && b["params"] is List) {
      attributeId = addIdsToAttributes(attributeId, b["params"]);
    }

    if (b.containsKey("properties") && b["properties"] is List) {
      attributeId = addIdsToAttributes(attributeId, b["properties"]);
    }

    return attributeId;
  }

  static int addIdsToAttributes(int attributeId, List attributes) {
    for (Map<String, Object> attribute in attributes) {
      attribute["id"] = attributeId;
      attributeId++;
    }
    return attributeId;
  }

  static void updateVersion1(Map json) {
    Map<String, int> actionToId = new Map<String, int>();
    Map<int, int> blockIdToAttributeIdOffset = new Map<int, int>();
    int attributeId = 0;

    if (!json.containsKey("blocks") || !(json["blocks"] is List)) {
      return;
    }

    for (Map<String, Object> b in json["blocks"]) {
      if (!b.containsKey("action")) { continue; }

      int id = actionToId.length;
      b["id"] = id;
      String action = b["action"];
      actionToId[action] = id;

      blockIdToAttributeIdOffset[id] = attributeId;

      attributeId = addIdsToParamsAndProps(attributeId, b);
    }

    if (!json.containsKey("program") || !(json["program"] is Map)) {
      return;
    }

    Map program = json["program"];
    if (!program.containsKey("chains") || program["chains"] is! List) {
      return;
    }

    for (List bs in program["chains"]) {
      for(Map b in bs) {
        addIdToBlock(actionToId, blockIdToAttributeIdOffset, b);
      }
    }
  }

  static void addIdToBlock(
    Map<String, int> actionToId,
    Map<int, int> blockIdToAttributeIdOffset,
    Map b
  ) {

    if (b.containsKey("action")) {
      String action = b["action"];
      if (actionToId.containsKey(action)) {
        int id = actionToId[action];
        b["id"] = id;
        addIdsToParamsAndProps(blockIdToAttributeIdOffset[id], b);
      }
    }

    if (b.containsKey("children") && b["children"] is List) {
      for (var child in b["children"]) {
        if (child is Map) {
          addIdToBlock(actionToId, blockIdToAttributeIdOffset, child);
        }
      }
    }

    if (b.containsKey("clauses") && b["clauses"] is List) {
      for (var clause in b["clauses"]) {
        if (clause is Map && clause.containsKey("children") && clause["children"] is List) {
          for (var child in clause['children']) {
            addIdToBlock(actionToId, blockIdToAttributeIdOffset, child);
          }
        }
      }
    }
  }

  static void addDisplayToSelectAttribute(Map attribute) {
    if (!attribute.containsKey("values") || attribute["values"] is! List) {
      return;
    }
    attribute["values"] = attribute["values"].map( (v) { return { "display": v, "actual": v }; } ).toList();
  }

  static void addDisplayToSelectAttributes(List attributes) {
    for (Map<String, Object> a in attributes.where( (f) => f.containsKey("type") && f["type"] == "select" )) {
      addDisplayToSelectAttribute(a);
    }
  }

  static void addSelectDisplayToBlock(Map b) {
    if (b.containsKey("params") && b["params"] is List) {
      addDisplayToSelectAttributes(b["params"]);
    }

    if (b.containsKey("properties") && b["properties"] is List) {
      addDisplayToSelectAttributes(b["properties"]);
    }

    if (b.containsKey("children") && b["children"] is List) {
      for (var child in b["children"]) {
        if (child is Map) {
          addSelectDisplayToBlock(child);
        }
      }
    }

    if (b.containsKey("clauses") && b["clauses"] is List) {
      for (var clause in b["clauses"]) {
        if (clause is Map && clause.containsKey("children") && clause["children"] is List) {
          for (var child in clause['children']) {
            addSelectDisplayToBlock(child);
          }
        }
      }
    }

  }

  static void updateVersion2(Map json) {
    if (!json.containsKey("blocks") || json["blocks"] is! List) {
      return;
    }

    for (Map<String, Object> b in json["blocks"]) {
      addSelectDisplayToBlock(b);
    }

    if (!json.containsKey("program") || json["program"] is! Map) {
      return;
    }

    Map program = json["program"];
    if (!program.containsKey("chains") || program["chains"] is! List) {
      return;
    }

    for (List bs in program["chains"]) {
      for(Map b in bs) {
        addSelectDisplayToBlock(b);
      }
    }
  }
}

File: src/versions/version-manager.ts
// NetTango Copyright (C) Michael S. Horn, Uri Wilensky, and Corey Brady. https://github.com/NetTango/NetTango

import {
  Attribute
, AttributeValue
, BlockDefinition
, BlockInstance
, CodeWorkspace
} from "../types/types"

import {
  makeAttributeDefault
, makeStringDefault
, makeNumberDefault
, makeExpressionDefault
} from "../blocks/attributes/attribute-factory"

import { Version1 } from "./version-1"
import { Version2 } from "./version-2"
import { Version3 } from "./version-3"
import { Version4 } from "./version-4"
import { Version5 } from "./version-5"
import { Version6 } from "./version-6"
import { ArrayUtils } from "../utils/array-utils"
import { ObjectUtils } from "../utils/object-utils"

class VersionManager {
  static readonly VERSION = 6

  static updateWorkspace(definition: any): CodeWorkspace {
    const version: number = definition.hasOwnProperty("version") ? definition.version : 0

    if (version > VersionManager.VERSION) {
      throw new Error(`Somehow the given model version (${version}) is greater than the supported NetTango version (${VersionManager.VERSION}).`)
    }

    const unvalidatedVersions = [Version1, Version2, Version3, Version4, Version5]
    unvalidatedVersions.slice(version).forEach( (v) => v.update(definition) )

    VersionManager.setBlockDefinitionIds(definition)

    if (version < 6) {
      definition.version = 5
      definition = Version5.validate(definition)
      definition = Version6.update(definition)
    }
    definition.version = VersionManager.VERSION

    const validated = Version6.validate(definition)

    VersionManager.fixupUntypedIssues(validated)

    return validated
  }

  static setBlockDefinitionIds(model: any): void {
    ObjectUtils.setIfUndefined(model, "blocks", [])

    if (model.blocks.length === 0) {
      return
    }

    // get the largest ID number among all block defs to get our next ID
    const ids = model.blocks.map( (b: any) => b.hasOwnProperty("id") && b.id !== null ? b.id : -1 )
    var nextId = Math.max(...ids) + 1

    // missing IDs, probably new blocks
    model.blocks.filter( (b: any) => !b.hasOwnProperty("id") || b.id === null || b.id === -1 ).forEach( (b: any) => {
      b.id = nextId
      nextId = nextId + 1
    })

    // duplicate IDs
    model.blocks.forEach( (b: any) => {
      const collisions = model.blocks.filter( (ob: any) => ob !== b && ob.id === b.id )
      collisions.forEach( (ob: any) => {
        ob.id = nextId
        nextId = nextId + 1
      })
    })

    model.blocks.forEach( (b: any) => ArrayUtils.maybeForEach(b, "params", (a: any, i: number) => {
      a.default = VersionManager.getAttributeDefaultValue(a)
      a.value = a.default
    }, (b: any) => b.params = []))

    model.blocks.forEach( (b: any) => ArrayUtils.maybeForEach(b, "properties", (a: any, i: number) => {
      a.default = VersionManager.getAttributeDefaultValue(a)
      a.value = a.default
    }, (b: any) => b.properties = []))

  }

  static getAttributeDefaultValue(a: any): string | number | null {
    if (a.hasOwnProperty("default")) {
      return a.default
    }

    switch (a.type) {
      case "int":
      case "range":
        return 10

      case "text":
      case "select":
        return ""

      case "num":
      case "bool":
        return "0"

      default:
        return null
    }
  }

  // Even though the data types are validated, the clauses and attributes also need to correspond
  // between definitions and instances, so we reset them here to be sure they are correct.
  // TODO: See if Zod can handle this validation as well.
  // -Jeremy B May 2021.
  static fixupUntypedIssues(model: CodeWorkspace) {
    const getDefById = (id: number): BlockDefinition | undefined =>
      model.blocks.filter( (b) => b.id === id )[0]

    const processInstance = (instance: BlockInstance) => {
      const definition = getDefById(instance.definitionId)
      VersionManager.resetBlockInstance(definition!, instance)
      instance.clauses.forEach( (clause) => {
        clause.blocks = clause.blocks.filter( (b) => getDefById(b.definitionId) !== undefined )
        clause.blocks.forEach(processInstance)
      })
    }

    model.program.chains.forEach( (chain) => {
      chain.blocks = chain.blocks.filter( (b) => getDefById(b.definitionId) !== undefined )
      chain.blocks.forEach(processInstance)
    })
  }

  static resetBlockInstance(definition: BlockDefinition, instance: BlockInstance) {
    instance.clauses = definition.clauses.map( (_, i) => {
      return instance.clauses.length > i ? instance.clauses[i] : { blocks: [] }
    })

    instance.params = definition.params.map( (a, i) => {
      return instance.params.length > i ? VersionManager.resetAttribute(a, instance.params[i]) : makeAttributeDefault(a)
    })

    instance.properties = definition.properties.map( (a, i) => {
      return instance.properties.length > i ? VersionManager.resetAttribute(a, instance.properties[i]) : makeAttributeDefault(a)
    })
  }

  static resetAttribute(attribute: Attribute, value: AttributeValue): AttributeValue {
    switch (attribute.type) {
      case "text":
      case "select":
        return (["text", "select"].includes(value.type) ? value : makeStringDefault(attribute))

      case "int":
      case "range":
        return (["int", "range"].includes(value.type) ? value : makeNumberDefault(attribute))

      case "num":
      case "bool":
        return (["num", "bool"].includes(value.type) ? value : makeExpressionDefault(attribute))
    }
  }

}

export { VersionManager }
